File: functions/src/agents/ide_guide/ide_guide_agent.ts
/**
 * IDE Guide Agent - Development Assistant for ClubRoyale
 * 
 * This agent accelerates development by providing:
 * - Code generation with project-aware context
 * - Architecture guidance and recommendations
 * - Bug analysis and fixing suggestions
 * - Feature implementation planning
 * 
 * Powered by Google GenKit + Gemini 2.0 Flash
 */

import { genkit, z } from 'genkit';
import { googleAI, gemini20FlashExp } from '@genkit-ai/googleai';
import { onCall, HttpsError } from 'firebase-functions/v2/https';

// Initialize GenKit with Gemini 2.0 Flash
const ai = genkit({
    plugins: [googleAI()],
    model: gemini20FlashExp,
});

// =============================================================================
// SCHEMAS
// =============================================================================

const CodeGenerationInputSchema = z.object({
    task: z.string().describe('The coding task to accomplish'),
    language: z.enum(['dart', 'typescript', 'python']).describe('Target language'),
    projectContext: z.object({
        existingPatterns: z.array(z.string()).optional().describe('Existing code patterns'),
        currentFile: z.string().optional().describe('Current file being edited'),
        dependencies: z.array(z.string()).optional().describe('Available dependencies'),
        architecture: z.string().optional().describe('Project architecture description'),
    }).optional(),
    codeStyle: z.object({
        naming: z.enum(['camelCase', 'snake_case', 'PascalCase']).optional(),
        useRiverpod: z.boolean().optional().describe('Use Riverpod for state management'),
        useFreezed: z.boolean().optional().describe('Use Freezed for data classes'),
    }).optional(),
});

const CodeGenerationOutputSchema = z.object({
    code: z.string().describe('Generated code'),
    explanation: z.string().describe('Explanation of the code'),
    filesToCreate: z.array(z.object({
        path: z.string(),
        content: z.string(),
        description: z.string(),
    })).optional().describe('New files to create'),
    filesToModify: z.array(z.object({
        path: z.string(),
        changes: z.string(),
        reason: z.string(),
    })).optional().describe('Existing files to modify'),
    tests: z.string().optional().describe('Unit tests for the code'),
    dependencies: z.array(z.string()).optional().describe('Required dependencies'),
});

const ArchitectureInputSchema = z.object({
    feature: z.string().describe('Feature to design'),
    requirements: z.array(z.string()).describe('Feature requirements'),
    constraints: z.array(z.string()).optional().describe('Technical constraints'),
    existingArchitecture: z.string().optional().describe('Current architecture description'),
});

const ArchitectureOutputSchema = z.object({
    recommendation: z.object({
        approach: z.string().describe('Recommended approach'),
        components: z.array(z.object({
            name: z.string(),
            type: z.enum(['service', 'widget', 'model', 'provider', 'screen', 'util']),
            responsibility: z.string(),
            file: z.string(),
        })).describe('Components to create'),
        dataFlow: z.string().describe('How data flows through components'),
        integrationPoints: z.array(z.string()).describe('Integration with existing code'),
    }),
    alternatives: z.array(z.object({
        approach: z.string(),
        pros: z.array(z.string()),
        cons: z.array(z.string()),
    })).optional().describe('Alternative approaches'),
    estimatedEffort: z.string().describe('Time estimate'),
    risks: z.array(z.string()).optional().describe('Potential risks'),
});

const BugAnalysisInputSchema = z.object({
    errorMessage: z.string().describe('The error message'),
    stackTrace: z.string().optional().describe('Stack trace if available'),
    codeContext: z.string().optional().describe('Relevant code snippet'),
    recentChanges: z.array(z.string()).optional().describe('Recent code changes'),
    environment: z.enum(['flutter_web', 'flutter_android', 'flutter_ios', 'functions']).optional(),
});

const BugAnalysisOutputSchema = z.object({
    rootCause: z.string().describe('Identified root cause'),
    explanation: z.string().describe('Why this error occurred'),
    suggestedFix: z.string().describe('Code fix suggestion'),
    preventionTips: z.array(z.string()).describe('How to prevent this in future'),
    relatedIssues: z.array(z.string()).optional().describe('Related issues to check'),
    confidence: z.number().min(0).max(1).describe('Confidence in diagnosis'),
});

const FeatureImplementationInputSchema = z.object({
    featureName: z.string().describe('Feature name'),
    featureDescription: z.string().describe('Detailed description'),
    userStories: z.array(z.string()).optional().describe('User stories'),
    acceptanceCriteria: z.array(z.string()).optional().describe('Acceptance criteria'),
    priority: z.enum(['low', 'medium', 'high', 'critical']).optional(),
});

const FeatureImplementationOutputSchema = z.object({
    implementationPlan: z.array(z.object({
        phase: z.number(),
        title: z.string(),
        description: z.string(),
        tasks: z.array(z.object({
            task: z.string(),
            file: z.string(),
            estimatedTime: z.string(),
            dependencies: z.array(z.string()).optional(),
            codeSnippet: z.string().optional(),
        })),
    })).describe('Step-by-step implementation plan'),
    totalEstimatedTime: z.string().describe('Total time estimate'),
    riskAreas: z.array(z.string()).describe('Areas that need careful attention'),
    testPlan: z.array(z.string()).describe('Testing approach'),
    filesAffected: z.array(z.string()).describe('Files that will be created or modified'),
});

// =============================================================================
// GENKIT FLOWS
// =============================================================================

/**
 * Code Generation Flow
 * Generates production-ready code following ClubRoyale patterns
 */
export const codeGenerationFlow = ai.defineFlow(
    {
        name: 'generateCode',
        inputSchema: CodeGenerationInputSchema,
        outputSchema: CodeGenerationOutputSchema,
    },
    async (input) => {
        const systemPrompt = `You are the IDE Guide Agent for ClubRoyale, a Flutter-based social gaming platform.

Project Stack:
- Flutter 3.x with Dart
- Firebase (Firestore, Auth, Functions, Storage)
- Riverpod for state management
- Freezed for immutable data classes
- GoRouter for navigation
- Google GenKit for AI features

Code Style:
- Use ${input.codeStyle?.naming || 'camelCase'} for variables
- ${input.codeStyle?.useRiverpod ? 'Use Riverpod providers' : 'Use standard patterns'}
- ${input.codeStyle?.useFreezed ? 'Use @freezed for data classes' : 'Use regular classes'}
- Always add documentation comments
- Use const constructors where possible
- Handle errors gracefully with try-catch

${input.projectContext?.architecture ? `Architecture: ${input.projectContext.architecture}` : ''}
${input.projectContext?.existingPatterns?.length ? `Existing Patterns: ${input.projectContext.existingPatterns.join(', ')}` : ''}
${input.projectContext?.dependencies?.length ? `Available Dependencies: ${input.projectContext.dependencies.join(', ')}` : ''}`;

        const prompt = `${systemPrompt}

Task: ${input.task}
Language: ${input.language}
${input.projectContext?.currentFile ? `Current File: ${input.projectContext.currentFile}` : ''}

Generate clean, production-ready code. Include:
1. The main code with proper documentation
2. Any additional files needed
3. Unit tests if applicable
4. Required dependencies`;

        const { output } = await ai.generate({
            prompt,
            output: { schema: CodeGenerationOutputSchema },
        });

        return output ?? { code: '', explanation: 'Failed to generate code' };
    }
);

/**
 * Architecture Guidance Flow
 * Provides architecture recommendations for new features
 */
export const architectureGuidanceFlow = ai.defineFlow(
    {
        name: 'architectureGuidance',
        inputSchema: ArchitectureInputSchema,
        outputSchema: ArchitectureOutputSchema,
    },
    async (input) => {
        const prompt = `You are the Architecture Advisor for ClubRoyale.

ClubRoyale Architecture Overview:
- lib/features/ - Feature modules (lobby, game, social, wallet, etc.)
- lib/core/ - Shared widgets, services, utilities
- lib/games/ - Game engines (call_break, marriage, teen_patti, in_between)
- lib/config/ - Theme and app configuration
- functions/src/ - Firebase Cloud Functions (TypeScript)

Each feature module contains:
- screens/ - UI screens
- widgets/ - Reusable widgets
- services/ - Business logic
- providers/ - Riverpod providers
- models/ - Data models

Feature: ${input.feature}
Requirements: 
${input.requirements.map((r, i) => `${i + 1}. ${r}`).join('\n')}
${input.constraints?.length ? `Constraints: ${input.constraints.join(', ')}` : ''}
${input.existingArchitecture ? `Current Architecture: ${input.existingArchitecture}` : ''}

Provide a detailed architecture recommendation including:
1. Recommended approach with justification
2. Components to create with responsibilities
3. Data flow diagram description
4. Integration points with existing code
5. Alternative approaches with pros/cons
6. Time estimate and risks`;

        const { output } = await ai.generate({
            prompt,
            output: { schema: ArchitectureOutputSchema },
        });

        return output ?? {
            recommendation: { approach: '', components: [], dataFlow: '', integrationPoints: [] },
            estimatedEffort: 'Unable to estimate',
        };
    }
);

/**
 * Bug Analysis Flow
 * Analyzes errors and suggests fixes
 */
export const bugAnalysisFlow = ai.defineFlow(
    {
        name: 'bugAnalysis',
        inputSchema: BugAnalysisInputSchema,
        outputSchema: BugAnalysisOutputSchema,
    },
    async (input) => {
        const prompt = `You are a senior Flutter/Firebase debugging expert for ClubRoyale.

Error: ${input.errorMessage}
${input.stackTrace ? `Stack Trace:\n${input.stackTrace}` : ''}
${input.codeContext ? `Code Context:\n${input.codeContext}` : ''}
${input.recentChanges?.length ? `Recent Changes: ${input.recentChanges.join(', ')}` : ''}
${input.environment ? `Environment: ${input.environment}` : ''}

Common ClubRoyale Issues to Consider:
- Riverpod provider scope issues
- Firestore permission errors
- WebRTC connection failures
- State management race conditions
- Widget lifecycle issues
- Null safety violations

Analyze this error and provide:
1. Root cause identification
2. Clear explanation of why it occurred
3. Specific code fix with example
4. Prevention tips for the future
5. Related issues that might occur`;

        const { output } = await ai.generate({
            prompt,
            output: { schema: BugAnalysisOutputSchema },
        });

        return output ?? {
            rootCause: 'Unable to analyze',
            explanation: 'Analysis failed',
            suggestedFix: '',
            preventionTips: [],
            confidence: 0,
        };
    }
);

/**
 * Feature Implementation Flow
 * Creates step-by-step implementation plans
 */
export const featureImplementationFlow = ai.defineFlow(
    {
        name: 'featureImplementation',
        inputSchema: FeatureImplementationInputSchema,
        outputSchema: FeatureImplementationOutputSchema,
    },
    async (input) => {
        const prompt = `You are a senior Flutter architect planning feature implementation for ClubRoyale.

Feature: ${input.featureName}
Description: ${input.featureDescription}
Priority: ${input.priority || 'medium'}

${input.userStories?.length ? `User Stories:\n${input.userStories.map((s, i) => `- ${s}`).join('\n')}` : ''}
${input.acceptanceCriteria?.length ? `Acceptance Criteria:\n${input.acceptanceCriteria.map((c, i) => `✓ ${c}`).join('\n')}` : ''}

ClubRoyale Project Structure:
- lib/features/{feature}/ - Feature modules
- lib/core/widgets/ - Shared widgets
- lib/core/services/ - Core services
- functions/src/ - Cloud functions

Create a detailed implementation plan with:
1. Phased approach (Backend → Service → UI → Polish)
2. Specific tasks with file paths
3. Time estimates per task
4. Dependencies between tasks
5. Code snippets for key components
6. Risk areas to watch
7. Testing strategy`;

        const { output } = await ai.generate({
            prompt,
            output: { schema: FeatureImplementationOutputSchema },
        });

        return output ?? {
            implementationPlan: [],
            totalEstimatedTime: 'Unable to estimate',
            riskAreas: [],
            testPlan: [],
            filesAffected: [],
        };
    }
);

// =============================================================================
// FIREBASE CALLABLE FUNCTIONS
// =============================================================================

/**
 * Generate Code - Firebase Callable Function
 */
export const generateCode = onCall(
    { maxInstances: 10, timeoutSeconds: 60 },
    async (request) => {
        if (!request.auth) {
            throw new HttpsError('unauthenticated', 'Authentication required');
        }

        try {
            const result = await codeGenerationFlow(request.data);
            return result;
        } catch (error: any) {
            console.error('Code generation error:', error);
            throw new HttpsError('internal', error.message);
        }
    }
);

/**
 * Architecture Guidance - Firebase Callable Function
 */
export const getArchitectureGuidance = onCall(
    { maxInstances: 10, timeoutSeconds: 60 },
    async (request) => {
        if (!request.auth) {
            throw new HttpsError('unauthenticated', 'Authentication required');
        }

        try {
            const result = await architectureGuidanceFlow(request.data);
            return result;
        } catch (error: any) {
            console.error('Architecture guidance error:', error);
            throw new HttpsError('internal', error.message);
        }
    }
);

/**
 * Bug Analysis - Firebase Callable Function
 */
export const analyzeBug = onCall(
    { maxInstances: 10, timeoutSeconds: 30 },
    async (request) => {
        if (!request.auth) {
            throw new HttpsError('unauthenticated', 'Authentication required');
        }

        try {
            const result = await bugAnalysisFlow(request.data);
            return result;
        } catch (error: any) {
            console.error('Bug analysis error:', error);
            throw new HttpsError('internal', error.message);
        }
    }
);

/**
 * Feature Implementation Plan - Firebase Callable Function
 */
export const planFeatureImplementation = onCall(
    { maxInstances: 10, timeoutSeconds: 60 },
    async (request) => {
        if (!request.auth) {
            throw new HttpsError('unauthenticated', 'Authentication required');
        }

        try {
            const result = await featureImplementationFlow(request.data);
            return result;
        } catch (error: any) {
            console.error('Feature planning error:', error);
            throw new HttpsError('internal', error.message);
        }
    }
);

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

/**
 * Quick code suggestion without full generation
 */
export const quickCodeSuggestionFlow = ai.defineFlow(
    {
        name: 'quickCodeSuggestion',
        inputSchema: z.object({
            prompt: z.string(),
            language: z.string(),
        }),
        outputSchema: z.object({
            suggestion: z.string(),
            confidence: z.number(),
        }),
    },
    async (input) => {
        const { output } = await ai.generate({
            prompt: `Quick ${input.language} code suggestion for ClubRoyale:\n${input.prompt}\n\nProvide a concise code snippet.`,
            output: {
                schema: z.object({
                    suggestion: z.string(),
                    confidence: z.number(),
                }),
            },
        });
        return output ?? { suggestion: '', confidence: 0 };
    }
);

/**
 * Explain code functionality
 */
export const explainCodeFlow = ai.defineFlow(
    {
        name: 'explainCode',
        inputSchema: z.object({
            code: z.string(),
            language: z.string(),
        }),
        outputSchema: z.object({
            explanation: z.string(),
            keyComponents: z.array(z.string()),
            potentialIssues: z.array(z.string()).optional(),
        }),
    },
    async (input) => {
        const { output } = await ai.generate({
            prompt: `Explain this ${input.language} code from ClubRoyale:\n\n${input.code}\n\nProvide a clear explanation with key components and any potential issues.`,
            output: {
                schema: z.object({
                    explanation: z.string(),
                    keyComponents: z.array(z.string()),
                    potentialIssues: z.array(z.string()).optional(),
                }),
            },
        });
        return output ?? { explanation: '', keyComponents: [] };
    }
);

File: lib/features/agents/services/ide_guide_agent.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:cloud_functions/cloud_functions.dart';

/// IDE Guide Agent provider
final ideGuideAgentProvider = Provider<IDEGuideAgent>((ref) {
  return IDEGuideAgent();
});

/// IDE Guide Agent - Development Assistant for ClubRoyale
/// 
/// Provides AI-powered development assistance:
/// - Code generation with project context
/// - Architecture guidance and recommendations
/// - Bug analysis and fixing suggestions
/// - Feature implementation planning
class IDEGuideAgent {
  final FirebaseFunctions _functions = FirebaseFunctions.instance;

  // ==================== CODE GENERATION ====================

  /// Generate code for a given task
  Future<CodeGenerationResult> generateCode({
    required String task,
    required String language,
    CodeContext? context,
    CodeStyle? style,
  }) async {
    try {
      final callable = _functions.httpsCallable('generateCode');
      final result = await callable.call<Map<String, dynamic>>({
        'task': task,
        'language': language,
        if (context != null) 'projectContext': context.toJson(),
        if (style != null) 'codeStyle': style.toJson(),
      });

      return CodeGenerationResult.fromJson(result.data);
    } catch (e) {
      return CodeGenerationResult(
        code: '// Error generating code: $e',
        explanation: 'Failed to generate code. Please try again.',
      );
    }
  }

  /// Quick code suggestion without full generation
  Future<String> quickSuggestion(String prompt, {String language = 'dart'}) async {
    try {
      final callable = _functions.httpsCallable('quickCodeSuggestion');
      final result = await callable.call<Map<String, dynamic>>({
        'prompt': prompt,
        'language': language,
      });
      return result.data['suggestion'] as String? ?? '';
    } catch (e) {
      return '// Error: $e';
    }
  }

  // ==================== ARCHITECTURE GUIDANCE ====================

  /// Get architecture recommendations for a feature
  Future<ArchitectureRecommendation> getArchitectureGuidance({
    required String feature,
    required List<String> requirements,
    List<String>? constraints,
    String? existingArchitecture,
  }) async {
    try {
      final callable = _functions.httpsCallable('getArchitectureGuidance');
      final result = await callable.call<Map<String, dynamic>>({
        'feature': feature,
        'requirements': requirements,
        if (constraints != null) 'constraints': constraints,
        if (existingArchitecture != null) 'existingArchitecture': existingArchitecture,
      });

      return ArchitectureRecommendation.fromJson(result.data);
    } catch (e) {
      return ArchitectureRecommendation(
        approach: 'Error getting recommendation',
        components: [],
        dataFlow: '',
        integrationPoints: [],
        estimatedEffort: 'Unknown',
      );
    }
  }

  // ==================== BUG ANALYSIS ====================

  /// Analyze an error and suggest fixes
  Future<BugAnalysisResult> analyzeBug({
    required String errorMessage,
    String? stackTrace,
    String? codeContext,
    List<String>? recentChanges,
    String? environment,
  }) async {
    try {
      final callable = _functions.httpsCallable('analyzeBug');
      final result = await callable.call<Map<String, dynamic>>({
        'errorMessage': errorMessage,
        if (stackTrace != null) 'stackTrace': stackTrace,
        if (codeContext != null) 'codeContext': codeContext,
        if (recentChanges != null) 'recentChanges': recentChanges,
        if (environment != null) 'environment': environment,
      });

      return BugAnalysisResult.fromJson(result.data);
    } catch (e) {
      return BugAnalysisResult(
        rootCause: 'Unable to analyze',
        explanation: 'Error analyzing bug: $e',
        suggestedFix: '',
        preventionTips: [],
        confidence: 0.0,
      );
    }
  }

  // ==================== FEATURE PLANNING ====================

  /// Create an implementation plan for a feature
  Future<FeatureImplementationPlan> planFeature({
    required String featureName,
    required String featureDescription,
    List<String>? userStories,
    List<String>? acceptanceCriteria,
    String priority = 'medium',
  }) async {
    try {
      final callable = _functions.httpsCallable('planFeatureImplementation');
      final result = await callable.call<Map<String, dynamic>>({
        'featureName': featureName,
        'featureDescription': featureDescription,
        if (userStories != null) 'userStories': userStories,
        if (acceptanceCriteria != null) 'acceptanceCriteria': acceptanceCriteria,
        'priority': priority,
      });

      return FeatureImplementationPlan.fromJson(result.data);
    } catch (e) {
      return FeatureImplementationPlan(
        phases: [],
        totalEstimatedTime: 'Unknown',
        riskAreas: ['Unable to generate plan'],
        testPlan: [],
        filesAffected: [],
      );
    }
  }

  /// Explain what a piece of code does
  Future<CodeExplanation> explainCode(String code, {String language = 'dart'}) async {
    try {
      final callable = _functions.httpsCallable('explainCode');
      final result = await callable.call<Map<String, dynamic>>({
        'code': code,
        'language': language,
      });

      return CodeExplanation(
        explanation: result.data['explanation'] as String? ?? '',
        keyComponents: List<String>.from(result.data['keyComponents'] ?? []),
        potentialIssues: List<String>.from(result.data['potentialIssues'] ?? []),
      );
    } catch (e) {
      return CodeExplanation(
        explanation: 'Error explaining code: $e',
        keyComponents: [],
        potentialIssues: [],
      );
    }
  }
}

// ==================== DATA MODELS ====================

class CodeContext {
  final List<String>? existingPatterns;
  final String? currentFile;
  final List<String>? dependencies;
  final String? architecture;

  CodeContext({
    this.existingPatterns,
    this.currentFile,
    this.dependencies,
    this.architecture,
  });

  Map<String, dynamic> toJson() => {
    if (existingPatterns != null) 'existingPatterns': existingPatterns,
    if (currentFile != null) 'currentFile': currentFile,
    if (dependencies != null) 'dependencies': dependencies,
    if (architecture != null) 'architecture': architecture,
  };
}

class CodeStyle {
  final String? naming;
  final bool useRiverpod;
  final bool useFreezed;

  CodeStyle({
    this.naming,
    this.useRiverpod = true,
    this.useFreezed = true,
  });

  Map<String, dynamic> toJson() => {
    if (naming != null) 'naming': naming,
    'useRiverpod': useRiverpod,
    'useFreezed': useFreezed,
  };
}

class CodeGenerationResult {
  final String code;
  final String explanation;
  final List<FileToCreate>? filesToCreate;
  final List<FileToModify>? filesToModify;
  final String? tests;
  final List<String>? dependencies;

  CodeGenerationResult({
    required this.code,
    required this.explanation,
    this.filesToCreate,
    this.filesToModify,
    this.tests,
    this.dependencies,
  });

  factory CodeGenerationResult.fromJson(Map<String, dynamic> json) {
    return CodeGenerationResult(
      code: json['code'] as String? ?? '',
      explanation: json['explanation'] as String? ?? '',
      filesToCreate: (json['filesToCreate'] as List<dynamic>?)
          ?.map((f) => FileToCreate.fromJson(f))
          .toList(),
      filesToModify: (json['filesToModify'] as List<dynamic>?)
          ?.map((f) => FileToModify.fromJson(f))
          .toList(),
      tests: json['tests'] as String?,
      dependencies: (json['dependencies'] as List<dynamic>?)?.cast<String>(),
    );
  }
}

class FileToCreate {
  final String path;
  final String content;
  final String description;

  FileToCreate({required this.path, required this.content, required this.description});

  factory FileToCreate.fromJson(Map<String, dynamic> json) {
    return FileToCreate(
      path: json['path'] as String? ?? '',
      content: json['content'] as String? ?? '',
      description: json['description'] as String? ?? '',
    );
  }
}

class FileToModify {
  final String path;
  final String changes;
  final String reason;

  FileToModify({required this.path, required this.changes, required this.reason});

  factory FileToModify.fromJson(Map<String, dynamic> json) {
    return FileToModify(
      path: json['path'] as String? ?? '',
      changes: json['changes'] as String? ?? '',
      reason: json['reason'] as String? ?? '',
    );
  }
}

class ArchitectureRecommendation {
  final String approach;
  final List<ComponentSpec> components;
  final String dataFlow;
  final List<String> integrationPoints;
  final List<AlternativeApproach>? alternatives;
  final String estimatedEffort;
  final List<String>? risks;

  ArchitectureRecommendation({
    required this.approach,
    required this.components,
    required this.dataFlow,
    required this.integrationPoints,
    this.alternatives,
    required this.estimatedEffort,
    this.risks,
  });

  factory ArchitectureRecommendation.fromJson(Map<String, dynamic> json) {
    final rec = json['recommendation'] as Map<String, dynamic>? ?? json;
    return ArchitectureRecommendation(
      approach: rec['approach'] as String? ?? '',
      components: (rec['components'] as List<dynamic>?)
          ?.map((c) => ComponentSpec.fromJson(c))
          .toList() ?? [],
      dataFlow: rec['dataFlow'] as String? ?? '',
      integrationPoints: (rec['integrationPoints'] as List<dynamic>?)?.cast<String>() ?? [],
      alternatives: (json['alternatives'] as List<dynamic>?)
          ?.map((a) => AlternativeApproach.fromJson(a))
          .toList(),
      estimatedEffort: json['estimatedEffort'] as String? ?? '',
      risks: (json['risks'] as List<dynamic>?)?.cast<String>(),
    );
  }
}

class ComponentSpec {
  final String name;
  final String type;
  final String responsibility;
  final String file;

  ComponentSpec({
    required this.name,
    required this.type,
    required this.responsibility,
    required this.file,
  });

  factory ComponentSpec.fromJson(Map<String, dynamic> json) {
    return ComponentSpec(
      name: json['name'] as String? ?? '',
      type: json['type'] as String? ?? '',
      responsibility: json['responsibility'] as String? ?? '',
      file: json['file'] as String? ?? '',
    );
  }
}

class AlternativeApproach {
  final String approach;
  final List<String> pros;
  final List<String> cons;

  AlternativeApproach({required this.approach, required this.pros, required this.cons});

  factory AlternativeApproach.fromJson(Map<String, dynamic> json) {
    return AlternativeApproach(
      approach: json['approach'] as String? ?? '',
      pros: (json['pros'] as List<dynamic>?)?.cast<String>() ?? [],
      cons: (json['cons'] as List<dynamic>?)?.cast<String>() ?? [],
    );
  }
}

class BugAnalysisResult {
  final String rootCause;
  final String explanation;
  final String suggestedFix;
  final List<String> preventionTips;
  final List<String>? relatedIssues;
  final double confidence;

  BugAnalysisResult({
    required this.rootCause,
    required this.explanation,
    required this.suggestedFix,
    required this.preventionTips,
    this.relatedIssues,
    required this.confidence,
  });

  factory BugAnalysisResult.fromJson(Map<String, dynamic> json) {
    return BugAnalysisResult(
      rootCause: json['rootCause'] as String? ?? '',
      explanation: json['explanation'] as String? ?? '',
      suggestedFix: json['suggestedFix'] as String? ?? '',
      preventionTips: (json['preventionTips'] as List<dynamic>?)?.cast<String>() ?? [],
      relatedIssues: (json['relatedIssues'] as List<dynamic>?)?.cast<String>(),
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class FeatureImplementationPlan {
  final List<ImplementationPhase> phases;
  final String totalEstimatedTime;
  final List<String> riskAreas;
  final List<String> testPlan;
  final List<String> filesAffected;

  FeatureImplementationPlan({
    required this.phases,
    required this.totalEstimatedTime,
    required this.riskAreas,
    required this.testPlan,
    required this.filesAffected,
  });

  factory FeatureImplementationPlan.fromJson(Map<String, dynamic> json) {
    return FeatureImplementationPlan(
      phases: (json['implementationPlan'] as List<dynamic>?)
          ?.map((p) => ImplementationPhase.fromJson(p))
          .toList() ?? [],
      totalEstimatedTime: json['totalEstimatedTime'] as String? ?? '',
      riskAreas: (json['riskAreas'] as List<dynamic>?)?.cast<String>() ?? [],
      testPlan: (json['testPlan'] as List<dynamic>?)?.cast<String>() ?? [],
      filesAffected: (json['filesAffected'] as List<dynamic>?)?.cast<String>() ?? [],
    );
  }
}

class ImplementationPhase {
  final int phase;
  final String title;
  final String description;
  final List<ImplementationTask> tasks;

  ImplementationPhase({
    required this.phase,
    required this.title,
    required this.description,
    required this.tasks,
  });

  factory ImplementationPhase.fromJson(Map<String, dynamic> json) {
    return ImplementationPhase(
      phase: json['phase'] as int? ?? 0,
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      tasks: (json['tasks'] as List<dynamic>?)
          ?.map((t) => ImplementationTask.fromJson(t))
          .toList() ?? [],
    );
  }
}

class ImplementationTask {
  final String task;
  final String file;
  final String estimatedTime;
  final List<String>? dependencies;
  final String? codeSnippet;

  ImplementationTask({
    required this.task,
    required this.file,
    required this.estimatedTime,
    this.dependencies,
    this.codeSnippet,
  });

  factory ImplementationTask.fromJson(Map<String, dynamic> json) {
    return ImplementationTask(
      task: json['task'] as String? ?? '',
      file: json['file'] as String? ?? '',
      estimatedTime: json['estimatedTime'] as String? ?? '',
      dependencies: (json['dependencies'] as List<dynamic>?)?.cast<String>(),
      codeSnippet: json['codeSnippet'] as String?,
    );
  }
}

class CodeExplanation {
  final String explanation;
  final List<String> keyComponents;
  final List<String> potentialIssues;

  CodeExplanation({
    required this.explanation,
    required this.keyComponents,
    required this.potentialIssues,
  });
}
